# Assignment #2
# DataFrames

# 1 and 2

In [3]:
import pandas as pd

In [4]:
president_polls = pd.read_csv('http://projects.fivethirtyeight.com/general-model/president_general_polls_2016.csv')

In [5]:
president_polls

cycle     branch        type                        matchup  \
0       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
1       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
2       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
3       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
4       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
...      ...        ...         ...                            ...   
12619   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12620   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12621   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12622   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12623   2016  President  polls-only  Clinton vs. Trump vs. Johnson   

      forecastdate          state   startdate    enddate  \
0          11/8/16           U.S.   11/3/2016  11/6/2016   
1          11/8/16           U.S.   11/1/2016  11/7/2016   
2          11/8/16           U.S.   11/2/2016  11/6/2016   
3          11/8/16           U.S.   11/4/2016  11/7/2016   
4          11/8/16           U.S.   11/3/2016  11/6/2016   
...            ...            ...         ...        ...   
12619      11/8/16  New Hampshire    7/9/2016  7/18/2016   
12620      11/8/16      Wisconsin  10/21/2016  11/2/2016   
12621      11/8/16       New York    8/7/2016  8/10/2016   
12622      11/8/16       Virginia   9/30/2016  10/6/2016   
12623      11/8/16      Wisconsin    6/9/2016  6/12/2016   

                          pollster grade  ...  adjpoll_clinton adjpoll_trump  \
0         ABC News/Washington Post    A+  ...         45.20163      41.72430   
1          Google Consumer Surveys     B  ...         43.34557      41.21439   
2                            Ipsos    A-  ...         42.02638      38.81620   
3                           YouGov     B  ...         45.65676      40.92004   
4                 Gravis Marketing    B-  ...         46.84089      42.33184   
...                            ...   ...  ...              ...           ...   
12619  University of New Hampshire    B+  ...         40.24983      43.04717   
12620                        Ipsos    A-  ...         46.54218      38.96884   
12621                Siena College     A  ...         53.83622      32.47939   
12622                        Ipsos    A-  ...         49.57558      39.96954   
12623         Marquette University     A  ...         46.40999      39.19839   

       adjpoll_johnson  adjpoll_mcmullin  multiversions  \
0             4.626221               NaN            NaN   
1             5.175792               NaN            NaN   
2             6.844734               NaN            NaN   
3             6.069454               NaN            NaN   
4             3.726098               NaN            NaN   
...                ...               ...            ...   
12619         6.924110               NaN            NaN   
12620              NaN               NaN            NaN   
12621         3.881193               NaN            NaN   
12622              NaN               NaN            NaN   
12623              NaN               NaN            NaN   

                                                     url  poll_id  \
0      https://www.washingtonpost.com/news/the-fix/wp...    48630   
1      https://datastudio.google.com/u/0/#/org//repor...    48847   
2      http://projects.fivethirtyeight.com/polls/2016...    48922   
3      https://d25d2506sfb94s.cloudfront.net/cumulus_...    48687   
4      http://www.gravispolls.com/2016/11/final-natio...    48848   
...                                                  ...      ...   
12619  https://cola.unh.edu/sites/cola.unh.edu/files/...    44650   
12620          http://www.reuters.com/statesofthenation/    48259   
12621  https://www.siena.edu/assets/files/news/SNY081...    44852   
12622          http://www.reuters.com/statesofthenation/    46675   
12623  https://l

# 1.

In [7]:
num_columns = president_polls.shape[1]

In [8]:
print(f"Number of columns: {num_columns}")

Number of columns: 27


# 2.

In [10]:
num_rows = president_polls.shape[0]

In [11]:
print(f"Number of rows: {num_rows}")

Number of rows: 12624


# 3. 

In [13]:
print("Column data types:")

Column data types:


In [14]:
print(president_polls.dtypes)

cycle                 int64
branch               object
type                 object
matchup              object
forecastdate         object
state                object
startdate            object
enddate              object
pollster             object
grade                object
samplesize          float64
population           object
poll_wt             float64
rawpoll_clinton     float64
rawpoll_trump       float64
rawpoll_johnson     float64
rawpoll_mcmullin    float64
adjpoll_clinton     float64
adjpoll_trump       float64
adjpoll_johnson     float64
adjpoll_mcmullin    float64
multiversions        object
url                  object
poll_id               int64
question_id           int64
createddate          object
timestamp            object
dtype: object


The column data types are mostly floats. I think the date columns could be changed to date times. As well as converting categorical columns from object to category. I believe both of these will improve the production of the dataset. 

# 4.

In [16]:
president_polls.describe()

cycle    samplesize       poll_wt  rawpoll_clinton  rawpoll_trump  \
count  12624.0  12621.000000  12624.000000     12624.000000   12624.000000   
mean    2016.0   1148.216068      0.255431        41.990860      39.828710   
std        0.0   2630.647790      0.669326         7.727792       7.877327   
min     2016.0     35.000000      0.000000        11.040000       4.000000   
25%     2016.0    447.000000      0.000358        38.000000      35.000000   
50%     2016.0    772.000000      0.008172        43.000000      40.000000   
75%     2016.0   1237.000000      0.086735        46.200000      45.000000   
max     2016.0  84292.000000      8.720654        88.000000      68.000000   

       rawpoll_johnson  rawpoll_mcmullin  adjpoll_clinton  adjpoll_trump  \
count      8397.000000         90.000000     12624.000000   12624.000000   
mean          7.381740         24.004000        43.323005      42.667719   
std           2.959274          5.638641         7.095235       6.948256   
min           0.000000          9.000000        17.064950       4.372936   
25%           5.400000         22.000000        40.210697      38.445588   
50%           7.000000         25.000000        44.148045      42.716265   
75%           9.000000         28.000000        46.908830      46.313520   
max          25.000000         31.000000        86.772180      72.437390   

       adjpoll_johnson  adjpoll_mcmullin       poll_id   question_id  
count      8397.000000         90.000000  12624.000000  12624.000000  
mean          4.655091         24.511497  45910.899477  71591.326046  
std           2.472636          5.174286   2864.536271   4699.401001  
min          -3.677883         11.028320  35362.000000  47244.000000  
25%           3.136837         22.814548  45151.750000  69548.500000  
50%           4.375392         25.136030  46384.500000  72516.500000  
75%           5.761469         28.067610  47741.250000  74724.250000  
max          20.366840         31.583810  48922.000000  76636.000000

# 5.

In [18]:
president_polls.nunique()

cycle                   1
branch                  1
type                    3
matchup                 1
forecastdate            1
state                  57
startdate             352
enddate               345
pollster              196
grade                  10
samplesize           1766
population              4
poll_wt              4399
rawpoll_clinton      1312
rawpoll_trump        1385
rawpoll_johnson       584
rawpoll_mcmullin       16
adjpoll_clinton     12569
adjpoll_trump       12582
adjpoll_johnson      6629
adjpoll_mcmullin       57
multiversions           1
url                  1304
poll_id              4208
question_id          4208
createddate           222
timestamp               3
dtype: int64

# 6.

In [20]:
president_polls.head(10) #see the first 10 results

cycle     branch        type                        matchup forecastdate  \
0   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
1   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
2   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
3   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
4   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
5   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
6   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
7   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
8   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   
9   2016  President  polls-plus  Clinton vs. Trump vs. Johnson      11/8/16   

        state  startdate    enddate  \
0        U.S.  11/3/2016  11/6/2016   
1        U.S.  11/1/2016  11/7/2016   
2        U.S.  11/2/2016  11/6/2016   
3        U.S.  11/4/2016  11/7/2016   
4        U.S.  11/3/2016  11/6/2016   
5        U.S.  11/3/2016  11/6/2016   
6        U.S.  11/2/2016  11/6/2016   
7        U.S.  11/3/2016  11/5/2016   
8  New Mexico  11/6/2016  11/6/2016   
9        U.S.  11/4/2016  11/7/2016   

                                            pollster grade  ...  \
0                           ABC News/Washington Post    A+  ...   
1                            Google Consumer Surveys     B  ...   
2                                              Ipsos    A-  ...   
3                                             YouGov     B  ...   
4                                   Gravis Marketing    B-  ...   
5  Fox News/Anderson Robbins Research/Shaw & Comp...     A  ...   
6                            CBS News/New York Times    A-  ...   
7                       NBC News/Wall Street Journal    A-  ...   
8                                           Zia Poll   NaN  ...   
9                                           IBD/TIPP    A-  ...   

   adjpoll_clinton adjpoll_trump  adjpoll_johnson  adjpoll_mcmullin  \
0         45.20163      41.72430         4.626221               NaN   
1         43.34557      41.21439         5.175792               NaN   
2         42.02638      38.81620         6.844734               NaN   
3         45.65676      40.92004         6.069454               NaN   
4         46.84089      42.33184         3.726098               NaN   
5         49.02208      43.95631         3.057876               NaN   
6         45.11649      40.92722         4.341786               NaN   
7         43.58576      40.77325         5.365788               NaN   
8         44.82594      41.59978         7.870127               NaN   
9         42.92745      42.23545         6.316175               NaN   

   multiversions                                                url  poll_id  \
0            NaN  https://www.washingtonpost.com/news/the-fix/wp...    48630   
1            NaN  https://datastudio.google.com/u/0/#/org//repor...    48847   
2            NaN  http://projects.fivethirtyeight.com/polls/2016...    48922   
3            NaN  https://d25d2506sfb94s.cloudfront.net/cumulus_...    48687   
4            NaN  http://www.gravispolls.com/2016/11/final-natio...    48848   
5            NaN  http://www.foxnews.com/politics/2016/11/07/fox...    48619   
6            NaN  http://www.cbsnews.com/news/cbs-news-poll-stat...    48521   
7            NaN  http://www.nbcnews.com/storyline/2016-election...    48480   
8            NaN  http://projects.fivethirtyeight.com.s3.amazona...    48614   
9            NaN  http://www.investors.com/politics/ibd-tipp-pre...    48916   

   question_id  createddate             timestamp  
0        76192      11/7/16  09:35:33  8 Nov 2016  
1        76443      11/7/16  09:35:33  8 Nov 2016  
2        76636      11/8/16  09:35:33  8 Nov 2016  
3        76262      11/7/16  09:35:33  8 Nov 2016  
4        76444      11/7/16  09:35:33  8 N

In [21]:
president_polls.tail(10) #see the last 10 results

cycle     branch        type                        matchup  \
12614   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12615   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12616   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12617   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12618   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12619   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12620   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12621   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12622   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12623   2016  President  polls-only  Clinton vs. Trump vs. Johnson   

      forecastdate           state   startdate    enddate  \
12614      11/8/16            Ohio    3/4/2016   3/6/2016   
12615      11/8/16  North Carolina   9/27/2016  10/2/2016   
12616      11/8/16         Florida   7/30/2016   8/7/2016   
12617      11/8/16            Ohio   6/30/2016  7/11/2016   
12618      11/8/16      New Jersey   9/30/2016  10/6/2016   
12619      11/8/16   New Hampshire    7/9/2016  7/18/2016   
12620      11/8/16       Wisconsin  10/21/2016  11/2/2016   
12621      11/8/16        New York    8/7/2016  8/10/2016   
12622      11/8/16        Virginia   9/30/2016  10/6/2016   
12623      11/8/16       Wisconsin    6/9/2016  6/12/2016   

                          pollster grade  ...  adjpoll_clinton adjpoll_trump  \
12614        Public Policy Polling    B+  ...         42.87491      41.37695   
12615        Quinnipiac University    A-  ...         47.72989      44.98194   
12616        Quinnipiac University    A-  ...         45.55449      46.36915   
12617        Quinnipiac University    A-  ...         40.71486      42.31798   
12618                        Ipsos    A-  ...         48.29913      35.40485   
12619  University of New Hampshire    B+  ...         40.24983      43.04717   
12620                        Ipsos    A-  ...         46.54218      38.96884   
12621                Siena College     A  ...         53.83622      32.47939   
12622                        Ipsos    A-  ...         49.57558      39.96954   
12623         Marquette University     A  ...         46.40999      39.19839   

       adjpoll_johnson  adjpoll_mcmullin  multiversions  \
12614              NaN               NaN            NaN   
12615         3.904715               NaN            NaN   
12616         3.338913               NaN            NaN   
12617         2.796137               NaN            NaN   
12618              NaN               NaN            NaN   
12619         6.924110               NaN            NaN   
12620              NaN               NaN            NaN   
12621         3.881193               NaN            NaN   
12622              NaN               NaN            NaN   
12623              NaN               NaN            NaN   

                                                     url  poll_id  \
12614  http://www.publicpolicypolling.com/pdf/2015/PP...    36554   
12615  https://www.qu.edu/news-and-events/quinnipiac-...    46382   
12616  https://drive.google.com/file/d/0B1XFXSmJVNknN...    44806   
12617  http://www.qu.edu/news-and-events/quinnipiac-u...    44566   
12618          http://www.reuters.com/statesofthenation/    46660   
12619  https://cola.unh.edu/sites/cola.unh.edu/files/...    44650   
12620          http://www.reuters.com/statesofthenation/    48259   
12621  https://www.siena.edu/assets/files/news/SNY081...    44852   
12622          http://www.reuters.com/statesofthenation/    46675   
12623  https://law.marquette.edu/poll/2016/06/15/new-...    44341   

       question_id  createddate             timestamp  
12614        50193      3/11/16  09:14:14  8 Nov 2016  
12615        72511      10/3/16  09:14:14  8 Nov 2016  
12616        68629       8/9/16  09:14:14  8 Nov 2016  
12617        67906      7/13/16  09:14:14  8 Nov 

From just printing the first and last 10 rows, I learned that this dataset is focused on the 2016 election as well as the branch type (president). In addition, it shows the poll type as well as the start and end date for each poll. It also shows the matchups and includes the states. Many pollsters are included and the grading of the pollsters as well. Also each poll is correlated with poll_id and question_id and there are URLs for further research. One potential concern is that there is some missing data for some candidates that just shows NaN. 

In [23]:
president_polls[['startdate','enddate']] #helps see when the polls started and ended

startdate    enddate
0       11/3/2016  11/6/2016
1       11/1/2016  11/7/2016
2       11/2/2016  11/6/2016
3       11/4/2016  11/7/2016
4       11/3/2016  11/6/2016
...           ...        ...
12619    7/9/2016  7/18/2016
12620  10/21/2016  11/2/2016
12621    8/7/2016  8/10/2016
12622   9/30/2016  10/6/2016
12623    6/9/2016  6/12/2016

[12624 rows x 2 columns]

In [24]:
president_polls[['adjpoll_clinton','adjpoll_trump','adjpoll_johnson','adjpoll_mcmullin']] #allows me to read data more clearly

adjpoll_clinton  adjpoll_trump  adjpoll_johnson  adjpoll_mcmullin
0             45.20163       41.72430         4.626221               NaN
1             43.34557       41.21439         5.175792               NaN
2             42.02638       38.81620         6.844734               NaN
3             45.65676       40.92004         6.069454               NaN
4             46.84089       42.33184         3.726098               NaN
...                ...            ...              ...               ...
12619         40.24983       43.04717         6.924110               NaN
12620         46.54218       38.96884              NaN               NaN
12621         53.83622       32.47939         3.881193               NaN
12622         49.57558       39.96954              NaN               NaN
12623         46.40999       39.19839              NaN               NaN

[12624 rows x 4 columns]

# 8.

In [26]:
print(president_polls.isnull().sum())

cycle                   0
branch                  0
type                    0
matchup                 0
forecastdate            0
state                   0
startdate               0
enddate                 0
pollster                0
grade                1287
samplesize              3
population              0
poll_wt                 0
rawpoll_clinton         0
rawpoll_trump           0
rawpoll_johnson      4227
rawpoll_mcmullin    12534
adjpoll_clinton         0
adjpoll_trump           0
adjpoll_johnson      4227
adjpoll_mcmullin    12534
multiversions       12588
url                     3
poll_id                 0
question_id             0
createddate             0
timestamp               0
dtype: int64


Yes this dataset is missing values. It is missing 1287 in grades. 3 in samplesize and url. As well as 4227 in rawpoll_johnson and adjpoll_johnson. 12534 are missing in rawpoll_mcmullin and adjpoll_mcmullin. 12588 multiversions are missing as well.

# 9.

In [29]:
president_polls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12624 entries, 0 to 12623
Data columns (total 27 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   cycle             12624 non-null  int64  
 1   branch            12624 non-null  object 
 2   type              12624 non-null  object 
 3   matchup           12624 non-null  object 
 4   forecastdate      12624 non-null  object 
 5   state             12624 non-null  object 
 6   startdate         12624 non-null  object 
 7   enddate           12624 non-null  object 
 8   pollster          12624 non-null  object 
 9   grade             11337 non-null  object 
 10  samplesize        12621 non-null  float64
 11  population        12624 non-null  object 
 12  poll_wt           12624 non-null  float64
 13  rawpoll_clinton   12624 non-null  float64
 14  rawpoll_trump     12624 non-null  float64
 15  rawpoll_johnson   8397 non-null   float64
 16  rawpoll_mcmullin  90 non-null     float6

This is not a clean dataset because we are missing data and values. We do not have complete data for #9,10,15,16,19,20,21,22. 

# 10.

The difference between this dataset and the mortality dataset that we used in class is the dataset in class was clean and a smaller dataset. This dataset is much larger and not clean.

# 11.

In [34]:
print(president_polls.columns)

Index(['cycle', 'branch', 'type', 'matchup', 'forecastdate', 'state',
       'startdate', 'enddate', 'pollster', 'grade', 'samplesize', 'population',
       'poll_wt', 'rawpoll_clinton', 'rawpoll_trump', 'rawpoll_johnson',
       'rawpoll_mcmullin', 'adjpoll_clinton', 'adjpoll_trump',
       'adjpoll_johnson', 'adjpoll_mcmullin', 'multiversions', 'url',
       'poll_id', 'question_id', 'createddate', 'timestamp'],
      dtype='object')


I think some of these columns need to be renamed by adding underscores in them.

In [36]:
president_polls.rename(columns = {'forecastdate': 'forecast_date', 'startdate':'start_date', 'enddate':'end_date','createddate':'created_date'}, inplace=True)

In [37]:
president_polls.tail()

cycle     branch        type                        matchup  \
12619   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12620   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12621   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12622   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12623   2016  President  polls-only  Clinton vs. Trump vs. Johnson   

      forecast_date          state  start_date   end_date  \
12619       11/8/16  New Hampshire    7/9/2016  7/18/2016   
12620       11/8/16      Wisconsin  10/21/2016  11/2/2016   
12621       11/8/16       New York    8/7/2016  8/10/2016   
12622       11/8/16       Virginia   9/30/2016  10/6/2016   
12623       11/8/16      Wisconsin    6/9/2016  6/12/2016   

                          pollster grade  ...  adjpoll_clinton adjpoll_trump  \
12619  University of New Hampshire    B+  ...         40.24983      43.04717   
12620                        Ipsos    A-  ...         46.54218      38.96884   
12621                Siena College     A  ...         53.83622      32.47939   
12622                        Ipsos    A-  ...         49.57558      39.96954   
12623         Marquette University     A  ...         46.40999      39.19839   

       adjpoll_johnson  adjpoll_mcmullin  multiversions  \
12619         6.924110               NaN            NaN   
12620              NaN               NaN            NaN   
12621         3.881193               NaN            NaN   
12622              NaN               NaN            NaN   
12623              NaN               NaN            NaN   

                                                     url  poll_id  \
12619  https://cola.unh.edu/sites/cola.unh.edu/files/...    44650   
12620          http://www.reuters.com/statesofthenation/    48259   
12621  https://www.siena.edu/assets/files/news/SNY081...    44852   
12622          http://www.reuters.com/statesofthenation/    46675   
12623  https://law.marquette.edu/poll/2016/06/15/new-...    44341   

       question_id  created_date             timestamp  
12619        68189       7/21/16  09:14:14  8 Nov 2016  
12620        75560       11/3/16  09:14:14  8 Nov 2016  
12621        68743       8/15/16  09:14:14  8 Nov 2016  
12622        72969      10/10/16  09:14:14  8 Nov 2016  
12623        66966       6/15/16  09:14:14  8 Nov 2016  

[5 rows x 27 columns]

In [38]:
president_polls.head()

cycle     branch        type                        matchup forecast_date  \
0   2016  President  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16   
1   2016  President  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16   
2   2016  President  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16   
3   2016  President  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16   
4   2016  President  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16   

  state start_date   end_date                  pollster grade  ...  \
0  U.S.  11/3/2016  11/6/2016  ABC News/Washington Post    A+  ...   
1  U.S.  11/1/2016  11/7/2016   Google Consumer Surveys     B  ...   
2  U.S.  11/2/2016  11/6/2016                     Ipsos    A-  ...   
3  U.S.  11/4/2016  11/7/2016                    YouGov     B  ...   
4  U.S.  11/3/2016  11/6/2016          Gravis Marketing    B-  ...   

   adjpoll_clinton adjpoll_trump  adjpoll_johnson  adjpoll_mcmullin  \
0         45.20163      41.72430         4.626221               NaN   
1         43.34557      41.21439         5.175792               NaN   
2         42.02638      38.81620         6.844734               NaN   
3         45.65676      40.92004         6.069454               NaN   
4         46.84089      42.33184         3.726098               NaN   

   multiversions                                                url  poll_id  \
0            NaN  https://www.washingtonpost.com/news/the-fix/wp...    48630   
1            NaN  https://datastudio.google.com/u/0/#/org//repor...    48847   
2            NaN  http://projects.fivethirtyeight.com/polls/2016...    48922   
3            NaN  https://d25d2506sfb94s.cloudfront.net/cumulus_...    48687   
4            NaN  http://www.gravispolls.com/2016/11/final-natio...    48848   

   question_id  created_date             timestamp  
0        76192       11/7/16  09:35:33  8 Nov 2016  
1        76443       11/7/16  09:35:33  8 Nov 2016  
2        76636       11/8/16  09:35:33  8 Nov 2016  
3        76262       11/7/16  09:35:33  8 Nov 2016  
4        76444       11/7/16  09:35:33  8 Nov 2016  

[5 rows x 27 columns]

# 12.

In [40]:
candidate_columns = ['rawpoll_clinton', 'rawpoll_trump', 'rawpoll_johnson', 'rawpoll_mcmullin',
                     'adjpoll_clinton', 'adjpoll_trump', 'adjpoll_johnson']

In [41]:
president_polls[candidate_columns].head()

rawpoll_clinton  rawpoll_trump  rawpoll_johnson  rawpoll_mcmullin  \
0            47.00          43.00             4.00               NaN   
1            38.03          35.69             5.46               NaN   
2            42.00          39.00             6.00               NaN   
3            45.00          41.00             5.00               NaN   
4            47.00          43.00             3.00               NaN   

   adjpoll_clinton  adjpoll_trump  adjpoll_johnson  
0         45.20163       41.72430         4.626221  
1         43.34557       41.21439         5.175792  
2         42.02638       38.81620         6.844734  
3         45.65676       40.92004         6.069454  
4         46.84089       42.33184         3.726098

In [42]:
president_polls[candidate_columns].tail()

rawpoll_clinton  rawpoll_trump  rawpoll_johnson  rawpoll_mcmullin  \
12619            37.00          37.00             10.0               NaN   
12620            47.48          38.96              NaN               NaN   
12621            50.00          25.00              9.0               NaN   
12622            50.49          39.05              NaN               NaN   
12623            46.00          37.00              NaN               NaN   

       adjpoll_clinton  adjpoll_trump  adjpoll_johnson  
12619         40.24983       43.04717         6.924110  
12620         46.54218       38.96884              NaN  
12621         53.83622       32.47939         3.881193  
12622         49.57558       39.96954              NaN  
12623         46.40999       39.19839              NaN

After studying this data, I have many takeaways. rawpoll_mcmullin is consistently missing (NaN) in all rows. I can conclude that McMullin wasn't included in as many polls as others. rawpoll_johnson is also missing values but fewer than McMullin. I can conclude that he wasn't included in as many either. I think that these values are missing and not as important because the main candidates were Trump and Clinton. Therefore the other 2 candidates were limited in their recognition. I think we should delete the columns about McMullin because he ran as independent and has the most missing data. I think save the ones with Johnson, Trump and Clinton. This will allow for the data to be more clean and easier to read.

# 13.

Remove branch because it is has the same value (President) for all rows.
Remove multiversions because it has missing values.
Remove created_data because it is not relevant for the results or for analyzing data.
Remove created_data and question_id because it does not add value to reading the polling data and it is not relevant for the results or for analyzing data.
Remove url because I don't they are a neccesity. They don't add value in the statistical aspect of reading the data.

In [46]:
columns_to_drop = ['branch','multiversions','created_date','question_id','poll_id', 'url']

In [47]:
president_polls_cleaned = president_polls.drop(columns=columns_to_drop)

In [48]:
president_polls_cleaned.head()

cycle        type                        matchup forecast_date state  \
0   2016  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16  U.S.   
1   2016  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16  U.S.   
2   2016  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16  U.S.   
3   2016  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16  U.S.   
4   2016  polls-plus  Clinton vs. Trump vs. Johnson       11/8/16  U.S.   

  start_date   end_date                  pollster grade  samplesize  ...  \
0  11/3/2016  11/6/2016  ABC News/Washington Post    A+      2220.0  ...   
1  11/1/2016  11/7/2016   Google Consumer Surveys     B     26574.0  ...   
2  11/2/2016  11/6/2016                     Ipsos    A-      2195.0  ...   
3  11/4/2016  11/7/2016                    YouGov     B      3677.0  ...   
4  11/3/2016  11/6/2016          Gravis Marketing    B-     16639.0  ...   

    poll_wt  rawpoll_clinton  rawpoll_trump  rawpoll_johnson  \
0  8.720654            47.00          43.00             4.00   
1  7.628472            38.03          35.69             5.46   
2  6.424334            42.00          39.00             6.00   
3  6.087135            45.00          41.00             5.00   
4  5.316449            47.00          43.00             3.00   

   rawpoll_mcmullin  adjpoll_clinton  adjpoll_trump  adjpoll_johnson  \
0               NaN         45.20163       41.72430         4.626221   
1               NaN         43.34557       41.21439         5.175792   
2               NaN         42.02638       38.81620         6.844734   
3               NaN         45.65676       40.92004         6.069454   
4               NaN         46.84089       42.33184         3.726098   

   adjpoll_mcmullin             timestamp  
0               NaN  09:35:33  8 Nov 2016  
1               NaN  09:35:33  8 Nov 2016  
2               NaN  09:35:33  8 Nov 2016  
3               NaN  09:35:33  8 Nov 2016  
4               NaN  09:35:33  8 Nov 2016  

[5 rows x 21 columns]

In class

In [50]:
president_polls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12624 entries, 0 to 12623
Data columns (total 27 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   cycle             12624 non-null  int64  
 1   branch            12624 non-null  object 
 2   type              12624 non-null  object 
 3   matchup           12624 non-null  object 
 4   forecast_date     12624 non-null  object 
 5   state             12624 non-null  object 
 6   start_date        12624 non-null  object 
 7   end_date          12624 non-null  object 
 8   pollster          12624 non-null  object 
 9   grade             11337 non-null  object 
 10  samplesize        12621 non-null  float64
 11  population        12624 non-null  object 
 12  poll_wt           12624 non-null  float64
 13  rawpoll_clinton   12624 non-null  float64
 14  rawpoll_trump     12624 non-null  float64
 15  rawpoll_johnson   8397 non-null   float64
 16  rawpoll_mcmullin  90 non-null     float6

In [51]:
president_polls.nunique()

cycle                   1
branch                  1
type                    3
matchup                 1
forecast_date           1
state                  57
start_date            352
end_date              345
pollster              196
grade                  10
samplesize           1766
population              4
poll_wt              4399
rawpoll_clinton      1312
rawpoll_trump        1385
rawpoll_johnson       584
rawpoll_mcmullin       16
adjpoll_clinton     12569
adjpoll_trump       12582
adjpoll_johnson      6629
adjpoll_mcmullin       57
multiversions           1
url                  1304
poll_id              4208
question_id          4208
created_date          222
timestamp               3
dtype: int64

In [52]:
president_polls.apply

<bound method DataFrame.apply of        cycle     branch        type                        matchup  \
0       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
1       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
2       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
3       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
4       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
...      ...        ...         ...                            ...   
12619   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12620   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12621   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12622   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12623   2016  President  polls-only  Clinton vs. Trump vs. Johnson   

      forecast_date          state  start_date   end_date  \
0           11/8/16           U.S.   11/3/2016  11/6/2016   
1   

In [53]:
president_polls.type.value_counts()

type
polls-plus    4208
now-cast      4208
polls-only    4208
Name: count, dtype: int64

any type of row that is not "now cast"

In [55]:
president_polls

cycle     branch        type                        matchup  \
0       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
1       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
2       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
3       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
4       2016  President  polls-plus  Clinton vs. Trump vs. Johnson   
...      ...        ...         ...                            ...   
12619   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12620   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12621   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12622   2016  President  polls-only  Clinton vs. Trump vs. Johnson   
12623   2016  President  polls-only  Clinton vs. Trump vs. Johnson   

      forecast_date          state  start_date   end_date  \
0           11/8/16           U.S.   11/3/2016  11/6/2016   
1           11/8/16           U.S.   11/1/2016  11/7/2016   
2           11/8/16           U.S.   11/2/2016  11/6/2016   
3           11/8/16           U.S.   11/4/2016  11/7/2016   
4           11/8/16           U.S.   11/3/2016  11/6/2016   
...             ...            ...         ...        ...   
12619       11/8/16  New Hampshire    7/9/2016  7/18/2016   
12620       11/8/16      Wisconsin  10/21/2016  11/2/2016   
12621       11/8/16       New York    8/7/2016  8/10/2016   
12622       11/8/16       Virginia   9/30/2016  10/6/2016   
12623       11/8/16      Wisconsin    6/9/2016  6/12/2016   

                          pollster grade  ...  adjpoll_clinton adjpoll_trump  \
0         ABC News/Washington Post    A+  ...         45.20163      41.72430   
1          Google Consumer Surveys     B  ...         43.34557      41.21439   
2                            Ipsos    A-  ...         42.02638      38.81620   
3                           YouGov     B  ...         45.65676      40.92004   
4                 Gravis Marketing    B-  ...         46.84089      42.33184   
...                            ...   ...  ...              ...           ...   
12619  University of New Hampshire    B+  ...         40.24983      43.04717   
12620                        Ipsos    A-  ...         46.54218      38.96884   
12621                Siena College     A  ...         53.83622      32.47939   
12622                        Ipsos    A-  ...         49.57558      39.96954   
12623         Marquette University     A  ...         46.40999      39.19839   

       adjpoll_johnson  adjpoll_mcmullin  multiversions  \
0             4.626221               NaN            NaN   
1             5.175792               NaN            NaN   
2             6.844734               NaN            NaN   
3             6.069454               NaN            NaN   
4             3.726098               NaN            NaN   
...                ...               ...            ...   
12619         6.924110               NaN            NaN   
12620              NaN               NaN            NaN   
12621         3.881193               NaN            NaN   
12622              NaN               NaN            NaN   
12623              NaN               NaN            NaN   

                                                     url  poll_id  \
0      https://www.washingtonpost.com/news/the-fix/wp...    48630   
1      https://datastudio.google.com/u/0/#/org//repor...    48847   
2      http://projects.fivethirtyeight.com/polls/2016...    48922   
3      https://d25d2506sfb94s.cloudfront.net/cumulus_...    48687   
4      http://www.gravispolls.com/2016/11/final-natio...    48848   
...                                                  ...      ...   
12619  https://cola.unh.edu/sites/cola.unh.edu/files/...    44650   
12620          http://www.reuters.com/statesofthenation/    48259   
12621  https://www.siena.edu/assets/files/news/SNY081...    44852   
12622          http://www.reuters.com/statesofthenation/    46675   
1262

# Assignment #4

In [56]:
president_polls.drop(columns=['cycle','matchup','forecast_date','timestamp','rawpoll_johnson','rawpoll_mcmullin','adjpoll_johnson','adjpoll_mcmullin','branch','multiversions','created_date','question_id','poll_id', 'url'], inplace=True)	

In [57]:
president_polls

type          state  start_date   end_date  \
0      polls-plus           U.S.   11/3/2016  11/6/2016   
1      polls-plus           U.S.   11/1/2016  11/7/2016   
2      polls-plus           U.S.   11/2/2016  11/6/2016   
3      polls-plus           U.S.   11/4/2016  11/7/2016   
4      polls-plus           U.S.   11/3/2016  11/6/2016   
...           ...            ...         ...        ...   
12619  polls-only  New Hampshire    7/9/2016  7/18/2016   
12620  polls-only      Wisconsin  10/21/2016  11/2/2016   
12621  polls-only       New York    8/7/2016  8/10/2016   
12622  polls-only       Virginia   9/30/2016  10/6/2016   
12623  polls-only      Wisconsin    6/9/2016  6/12/2016   

                          pollster grade  samplesize population   poll_wt  \
0         ABC News/Washington Post    A+      2220.0         lv  8.720654   
1          Google Consumer Surveys     B     26574.0         lv  7.628472   
2                            Ipsos    A-      2195.0         lv  6.424334   
3                           YouGov     B      3677.0         lv  6.087135   
4                 Gravis Marketing    B-     16639.0         rv  5.316449   
...                            ...   ...         ...        ...       ...   
12619  University of New Hampshire    B+       469.0         lv  0.000000   
12620                        Ipsos    A-       653.0         lv  0.000000   
12621                Siena College     A       717.0         rv  0.000000   
12622                        Ipsos    A-       479.0         lv  0.000000   
12623         Marquette University     A       666.0         lv  0.000000   

       rawpoll_clinton  rawpoll_trump  adjpoll_clinton  adjpoll_trump  
0                47.00          43.00         45.20163       41.72430  
1                38.03          35.69         43.34557       41.21439  
2                42.00          39.00         42.02638       38.81620  
3                45.00          41.00         45.65676       40.92004  
4                47.00          43.00         46.84089       42.33184  
...                ...            ...              ...            ...  
12619            37.00          37.00         40.24983       43.04717  
12620            47.48          38.96         46.54218       38.96884  
12621            50.00          25.00         53.83622       32.47939  
12622            50.49          39.05         49.57558       39.96954  
12623            46.00          37.00         46.40999       39.19839  

[12624 rows x 13 columns]

In [58]:
president_polls = president_polls[president_polls.type == 'now-cast']

In [59]:
president_polls

type           state  start_date   end_date  \
4208  now-cast            U.S.   11/3/2016  11/6/2016   
4209  now-cast            U.S.   11/1/2016  11/7/2016   
4210  now-cast            U.S.   11/2/2016  11/6/2016   
4211  now-cast            U.S.   11/4/2016  11/7/2016   
4212  now-cast            U.S.   11/3/2016  11/6/2016   
...        ...             ...         ...        ...   
8411  now-cast  North Carolina   8/26/2016  8/27/2016   
8412  now-cast  North Carolina    3/9/2016  3/10/2016   
8413  now-cast  North Carolina   7/31/2016   8/2/2016   
8414  now-cast            Utah    8/5/2016  8/25/2016   
8415  now-cast        Missouri  10/21/2016  11/2/2016   

                      pollster grade  samplesize population   poll_wt  \
4208  ABC News/Washington Post    A+      2220.0         lv  8.720654   
4209   Google Consumer Surveys     B     26574.0         lv  7.628472   
4210                     Ipsos    A-      2195.0         lv  6.424334   
4211                    YouGov     B      3677.0         lv  6.087135   
4212          Gravis Marketing    B-     16639.0         rv  5.316449   
...                        ...   ...         ...        ...       ...   
8411     Public Policy Polling    B+      1177.0         lv  0.000000   
8412                 SurveyUSA     A      1576.0         lv  0.000000   
8413                 SurveyUSA     A       400.0         lv  0.000000   
8414                     Ipsos    A-       177.0         lv  0.000000   
8415                     Ipsos    A-       405.0         lv  0.000000   

      rawpoll_clinton  rawpoll_trump  adjpoll_clinton  adjpoll_trump  
4208            47.00          43.00         45.23046       41.68534  
4209            38.03          35.69         43.36603       41.20737  
4210            42.00          39.00         42.05478       38.81421  
4211            45.00          41.00         45.60811       40.89765  
4212            47.00          43.00         46.93448       42.38145  
...               ...            ...              ...            ...  
8411            45.00          44.00         44.40263       44.69241  
8412            42.00          49.00         38.91546       48.83620  
8413            42.00          46.00         42.90049       48.56213  
8414            33.67          38.53         33.17161       40.04859  
8415            41.09          47.19         39.82209       46.82893  

[4208 rows x 13 columns]

1. Do we need both the adjpoll and raw data for Clinton and Trump?  I think everyone should consider keeping either the adjpoll or the raw data columns.  Which column do you think you should remove?  Why?  Explain your answer then move forward with deleting those columns.
- I think we should remove the raw data columns and keep the adjusted results because they account for various factors such as polling bias and adjustments and they are also more reliable. 

In [61]:
president_polls.drop(columns=['rawpoll_clinton','rawpoll_trump'], inplace=True)

/var/folders/kr/qsc4lnv92nnfbz7q4w4m_j400000gn/T/ipykernel_6551/1418371706.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  president_polls.drop(columns=['rawpoll_clinton','rawpoll_trump'], inplace=True)


In [62]:
president_polls

type           state  start_date   end_date  \
4208  now-cast            U.S.   11/3/2016  11/6/2016   
4209  now-cast            U.S.   11/1/2016  11/7/2016   
4210  now-cast            U.S.   11/2/2016  11/6/2016   
4211  now-cast            U.S.   11/4/2016  11/7/2016   
4212  now-cast            U.S.   11/3/2016  11/6/2016   
...        ...             ...         ...        ...   
8411  now-cast  North Carolina   8/26/2016  8/27/2016   
8412  now-cast  North Carolina    3/9/2016  3/10/2016   
8413  now-cast  North Carolina   7/31/2016   8/2/2016   
8414  now-cast            Utah    8/5/2016  8/25/2016   
8415  now-cast        Missouri  10/21/2016  11/2/2016   

                      pollster grade  samplesize population   poll_wt  \
4208  ABC News/Washington Post    A+      2220.0         lv  8.720654   
4209   Google Consumer Surveys     B     26574.0         lv  7.628472   
4210                     Ipsos    A-      2195.0         lv  6.424334   
4211                    YouGov     B      3677.0         lv  6.087135   
4212          Gravis Marketing    B-     16639.0         rv  5.316449   
...                        ...   ...         ...        ...       ...   
8411     Public Policy Polling    B+      1177.0         lv  0.000000   
8412                 SurveyUSA     A      1576.0         lv  0.000000   
8413                 SurveyUSA     A       400.0         lv  0.000000   
8414                     Ipsos    A-       177.0         lv  0.000000   
8415                     Ipsos    A-       405.0         lv  0.000000   

      adjpoll_clinton  adjpoll_trump  
4208         45.23046       41.68534  
4209         43.36603       41.20737  
4210         42.05478       38.81421  
4211         45.60811       40.89765  
4212         46.93448       42.38145  
...               ...            ...  
8411         44.40263       44.69241  
8412         38.91546       48.83620  
8413         42.90049       48.56213  
8414         33.17161       40.04859  
8415         39.82209       46.82893  

[4208 rows x 11 columns]

2. Review the “type” column, determine if it is still needed to analyze this dataset.   If you do not, delete that column as well.
- I don't think the "type" column is needed to analyze this dataset because it is already filtered for "now-cast" and it no longer has any use to it.

In [64]:
president_polls.drop(columns=['type'], inplace=True)

/var/folders/kr/qsc4lnv92nnfbz7q4w4m_j400000gn/T/ipykernel_6551/191796290.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  president_polls.drop(columns=['type'], inplace=True)


In [65]:
president_polls

state  start_date   end_date                  pollster grade  \
4208            U.S.   11/3/2016  11/6/2016  ABC News/Washington Post    A+   
4209            U.S.   11/1/2016  11/7/2016   Google Consumer Surveys     B   
4210            U.S.   11/2/2016  11/6/2016                     Ipsos    A-   
4211            U.S.   11/4/2016  11/7/2016                    YouGov     B   
4212            U.S.   11/3/2016  11/6/2016          Gravis Marketing    B-   
...              ...         ...        ...                       ...   ...   
8411  North Carolina   8/26/2016  8/27/2016     Public Policy Polling    B+   
8412  North Carolina    3/9/2016  3/10/2016                 SurveyUSA     A   
8413  North Carolina   7/31/2016   8/2/2016                 SurveyUSA     A   
8414            Utah    8/5/2016  8/25/2016                     Ipsos    A-   
8415        Missouri  10/21/2016  11/2/2016                     Ipsos    A-   

      samplesize population   poll_wt  adjpoll_clinton  adjpoll_trump  
4208      2220.0         lv  8.720654         45.23046       41.68534  
4209     26574.0         lv  7.628472         43.36603       41.20737  
4210      2195.0         lv  6.424334         42.05478       38.81421  
4211      3677.0         lv  6.087135         45.60811       40.89765  
4212     16639.0         rv  5.316449         46.93448       42.38145  
...          ...        ...       ...              ...            ...  
8411      1177.0         lv  0.000000         44.40263       44.69241  
8412      1576.0         lv  0.000000         38.91546       48.83620  
8413       400.0         lv  0.000000         42.90049       48.56213  
8414       177.0         lv  0.000000         33.17161       40.04859  
8415       405.0         lv  0.000000         39.82209       46.82893  

[4208 rows x 10 columns]

3. Run value counts on the “Population” column to identify which values are in the column.


In [112]:
president_polls['population'].value_counts()

population
lv    3727
rv     418
v       42
a       21
Name: count, dtype: int64

4. Based on the instructions, you were informed that you ONLY need to keep the likely voters and registered voter information in the “Population” column.  Move forward with removing “a” and “v” columns. 


In [135]:
president_polls.query('population in ["lv", "rv"]')

state  start_date   end_date                  pollster grade  \
4208            U.S.   11/3/2016  11/6/2016  ABC News/Washington Post    A+   
4209            U.S.   11/1/2016  11/7/2016   Google Consumer Surveys     B   
4210            U.S.   11/2/2016  11/6/2016                     Ipsos    A-   
4211            U.S.   11/4/2016  11/7/2016                    YouGov     B   
4212            U.S.   11/3/2016  11/6/2016          Gravis Marketing    B-   
...              ...         ...        ...                       ...   ...   
8411  North Carolina   8/26/2016  8/27/2016     Public Policy Polling    B+   
8412  North Carolina    3/9/2016  3/10/2016                 SurveyUSA     A   
8413  North Carolina   7/31/2016   8/2/2016                 SurveyUSA     A   
8414            Utah    8/5/2016  8/25/2016                     Ipsos    A-   
8415        Missouri  10/21/2016  11/2/2016                     Ipsos    A-   

      samplesize population   poll_wt  adjpoll_clinton  adjpoll_trump  
4208      2220.0         lv  8.720654         45.23046       41.68534  
4209     26574.0         lv  7.628472         43.36603       41.20737  
4210      2195.0         lv  6.424334         42.05478       38.81421  
4211      3677.0         lv  6.087135         45.60811       40.89765  
4212     16639.0         rv  5.316449         46.93448       42.38145  
...          ...        ...       ...              ...            ...  
8411      1177.0         lv  0.000000         44.40263       44.69241  
8412      1576.0         lv  0.000000         38.91546       48.83620  
8413       400.0         lv  0.000000         42.90049       48.56213  
8414       177.0         lv  0.000000         33.17161       40.04859  
8415       405.0         lv  0.000000         39.82209       46.82893  

[4145 rows x 10 columns]

5. Review the columns in your dataset, is it necessary to rename some of the columns to align with the data elements?  If so, rename the columns that you think should change.
- I am changing samplesize to sample_size, adjpoll_trump to adjusted_trump, and adjpoll_clinton to adjusted_clinton.

In [140]:
president_polls.rename(columns={'adjpoll_clinton': 'adjusted_clinton', 'adjpoll_trump': 'adjusted_trump', 'samplesize':'sample_size'})

state  start_date   end_date                  pollster grade  \
4208            U.S.   11/3/2016  11/6/2016  ABC News/Washington Post    A+   
4209            U.S.   11/1/2016  11/7/2016   Google Consumer Surveys     B   
4210            U.S.   11/2/2016  11/6/2016                     Ipsos    A-   
4211            U.S.   11/4/2016  11/7/2016                    YouGov     B   
4212            U.S.   11/3/2016  11/6/2016          Gravis Marketing    B-   
...              ...         ...        ...                       ...   ...   
8411  North Carolina   8/26/2016  8/27/2016     Public Policy Polling    B+   
8412  North Carolina    3/9/2016  3/10/2016                 SurveyUSA     A   
8413  North Carolina   7/31/2016   8/2/2016                 SurveyUSA     A   
8414            Utah    8/5/2016  8/25/2016                     Ipsos    A-   
8415        Missouri  10/21/2016  11/2/2016                     Ipsos    A-   

      sample_size population   poll_wt  adjusted_clinton  adjusted_trump  
4208       2220.0         lv  8.720654          45.23046        41.68534  
4209      26574.0         lv  7.628472          43.36603        41.20737  
4210       2195.0         lv  6.424334          42.05478        38.81421  
4211       3677.0         lv  6.087135          45.60811        40.89765  
4212      16639.0         rv  5.316449          46.93448        42.38145  
...           ...        ...       ...               ...             ...  
8411       1177.0         lv  0.000000          44.40263        44.69241  
8412       1576.0         lv  0.000000          38.91546        48.83620  
8413        400.0         lv  0.000000          42.90049        48.56213  
8414        177.0         lv  0.000000          33.17161        40.04859  
8415        405.0         lv  0.000000          39.82209        46.82893  

[4208 rows x 10 columns]